runner/ollamarunner/runner.go

```diff
@@ -338,7 +338,14 @@ func (s *Server) processBatch() error {
 				if len(seq.pendingInputs) == 0 {
 					err := s.cache.ShiftCacheSlot(seq.cache, seq.numKeep)
 					if err != nil {
-						return err
+						var reprocess *ErrReprocessInputs
+						if errors.As(err, &reprocess) {
+							// Prepend these inputs to the sequence's inputs queue for reprocessing
+							seq.inputs = append(reprocess.Inputs, seq.inputs...)
```

Comment:
This loop is currently iterating over `seq.inputs` so I don't think that this will do the right thing, we will likely get some of the existing inputs processed before the ones we want to prepend.

We could just do `input := seq.inputs[i]` after the`if len(seq.cache.Inputs)+len(seq.pendingInputs)+1 > s.cache.numCtx {` block. We only shift if we have no pending inputs, so that makes things simpler.

Reply:
Good catch. I've implemented a slightly different fix, returning early after prepending the reprocess inputs.

This ensures we'll process inputs in the correct order when the function is called again. Let me know if you prefer your suggested approach of moving the input selection.

Reply:
We might be building a batch with inputs from multiple sequences. If we have already gone through other sequences and return early here then the tokens we have already added to the batch will be lost.

Reply:
Changed this to continue, I believe it handles the multiple sequences by not returning while processing